### PR TITLE
0.13.1 update gas costs + events cost tracking

### LINKED ISSUE #1796: Include event data storage costs in gas estimations
### Which components does the task require to be changed? (think hard pls)

snforge

### Description

We currently don't count the storage cost of events and their data, 
it should be included in overall costs calculation

It's tagged in the code with a TODO - please remember to remove it after addressing this issue

### LINKED ISSUE #1821: Change gas cost constants to 0.13.1 starknet
### Which components does the task require to be changed? (think hard pls)

snforge

### Description

We should swap them so that they are compliant with the new starknet version

### PR DESCRIPTION
**Pull request by Arcticae**
Closes #1821
Closes #1796

## Introduced changes

- Aligns the costs with 0.13.1 version
- Adds event cost tracking

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs [hunk 1]
```diff
@@ -868,8 +870,15 @@ pub fn get_all_used_resources(runtime: ForgeRuntime) -> UsedResources {
 
     let execution_resources = top_call.borrow().used_execution_resources.clone();
     let top_call_syscalls = top_call.borrow().used_syscalls.clone();
+    let events = runtime_call_info
+        .execution
+        .events
+        .iter()
+        .map(|ordered_event| ordered_event.event.clone())
+        .collect();
```
Comment (piotmag769) [inline crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:878]: Is there no method for that? It seems like we are ignoring events from inner calls (unless events are like resources not like messages in the sense that they include the events from inner calls - then the name `top_call_events` above is missleading)

File: crates/forge-runner/src/gas.rs [hunk 2]
```diff
@@ -37,11 +39,30 @@ pub fn calculate_used_gas(
...
 
-    let gas = l1_and_vm_costs + messaging_gas_vector;
+    let events_costs = get_events_cost(&resources.events, transaction_context);
+
+    let gas = l1_and_vm_costs + messaging_gas_vector + events_costs;
 
     Ok(gas.l1_gas)
 }
 
+fn get_events_cost(events: &[EventContent], transaction_context: &TransactionContext) -> GasVector {
```
Comment (piotmag769) [inline crates/forge-runner/src/gas.rs:49]: We can pass a vector and consume it so we get rid of the clone. Just unpack the struct in the previous function to satisfy the borrow checker
```rs
let UsedResources {events, a, b, c} = used_resources;
```

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs (hunk 1 above)
Comment (Arcticae) [reply crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs]: My mistake, i added a test for this. Was missing the inner events, which are now collected

File: crates/forge-runner/src/gas.rs (hunk 2 above)
Comment (Arcticae) [reply crates/forge-runner/src/gas.rs]: Done

File: crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs [hunk 3]
```diff
@@ -870,8 +874,20 @@ pub fn get_all_used_resources(runtime: ForgeRuntime) -> UsedResources {
...
+        .into_iter() // This method iterates over inner calls as well
+        .map(|call_info| {
+            call_info
+                .execution
+                .events
+                .iter()
+                .map(|evt| evt.event.clone())
+                .collect::<Vec<EventContent>>()
+        })
+        .fold(vec![], |a, b| chain!(a, b).collect());
```
Comment (Draggu) [inline crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:887]:
```suggestion
        .flat_map(|call_info| {
            call_info
                .execution
                .events
                .iter()
                .map(|evt| evt.event.clone())
        })
        .collect::<Vec<EventContent>>();
```

Comment (Draggu) [reply crates/cheatnet/src/runtime_extensions/forge_runtime_extension/mod.rs:887]: Cleaner and faster

File: crates/forge/tests/integration/gas.rs [hunk 4]
```diff
@@ -675,7 +675,76 @@ fn l1_handler_cost() {
...
+            fn event_emission_cost() {
+                let contract_address = declare("GasChecker").deploy(@array![]).unwrap();
+                let dispatcher = IGasCheckerDispatcher { contract_address };
+
+                dispatcher.emit_event(50);
+            }
+        "#
+        ),
+        Contract::from_code_path(
+            "GasChecker".to_string(),
```
Comment (Draggu) [inline crates/forge/tests/integration/gas.rs:736]:
```suggestion
            "GasChecker",
```

**Review by Draggu — APPROVED** [review]

File: crates/forge-runner/src/gas.rs [hunk 5]
```diff
@@ -39,11 +41,33 @@ pub fn calculate_used_gas(
...
+        .into_iter()
+        .map(|content| OrderedEvent {
+            order: 0, // Order does not matter here
+            event: content,
+        })
+        .collect();
+    let miligas = get_events_milligas_cost(&ordered_events, versioned_constants);
+    GasVector {
+        l1_gas: miligas / 1000_u128,
+        l1_data_gas: 0_u128,
```
Comment (drknzz) [inline crates/forge-runner/src/gas.rs:67]:
```suggestion
    let milligas = get_events_milligas_cost(&ordered_events, versioned_constants);
    GasVector {
        l1_gas: milligas / 1000_u128,
        l1_data_gas: 0_u128,
```

File: crates/forge/tests/e2e/steps.rs [hunk 6]
```diff
@@ -73,16 +73,16 @@ fn should_allow_more_than_4kk() {
...
+                [PASS] steps::tests::steps_3999987 [..]
                 [PASS] steps::tests::steps_5700031 [..]
                 Tests: 4 passed, 0 failed, 0 skipped, 0 ignored, 0 filtered out
             "
         ),
     );
 }
 #[test]
-fn should_default_to_3kk() {
+fn should_default_to_4kk() {
```
Comment (drknzz) [inline crates/forge/tests/e2e/steps.rs:85]:
```suggestion
fn should_default_to_4m() {
```

**Review by drknzz — APPROVED** [review]
giga good

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**